src/Planner/PlannerJoins.cpp

```diff
@@ -385,6 +600,203 @@ void buildJoinClause(
     }
 }
 
+JoinClauses buildJoinClauses(
+    ActionsDAG & left_dag,
+    ActionsDAG & right_dag,
+    ActionsDAG & joined_dag,
+    const PlannerContextPtr & planner_context,
+    const QueryTreeNodePtr & join_expression,
+    const TableExpressionSet & left_table_expressions,
+    const TableExpressionSet & right_table_expressions,
+    const JoinNode & join_node)
+{
+    if (join_expression->getNodeType() != QueryTreeNodeType::FUNCTION)
+    {
+        JoinClauses result;
+        result.emplace_back();
+        buildSimpleJoinClause(
+            left_dag,
+            right_dag,
+            joined_dag,
+            planner_context,
+            join_expression,
+            left_table_expressions,
+            right_table_expressions,
+            join_node,
+            result.front());
+        return result;
+    }
+
+    std::unordered_map<IQueryTreeNode *, JoinClauses> built_clauses;
+    std::stack<QueryTreeNodePtr> nodes_to_process;
+    nodes_to_process.push(join_expression);
+
+    while (!nodes_to_process.empty())
```

Comment:
Can we say that the algorithm below essentially builds a DNF by combining expressions into terms, as long as there is no `==` function inside? Could you confirm or point out how it differs?